python/ray/ml/config.py

```diff
@@ -1,13 +1,65 @@
 from dataclasses import dataclass
-from typing import Any, Dict, List, Optional
+from typing import Dict, Any, Optional, List, TYPE_CHECKING
 
-from ray.tune.callback import Callback
 from ray.util import PublicAPI
 
+if TYPE_CHECKING:
+    from ray.tune.trainable import PlacementGroupFactory
+    from ray.tune.callback import Callback
+
+
 ScalingConfig = Dict[str, Any]
 
 
 @dataclass
+class ScalingConfigDataClass:
+    """Configuration for scaling training.
+
+    This is the schema for the scaling_config dict, and after beta, this will be the
+    actual representation for Scaling config objects.
+
+    num_workers: The number of workers (Ray actors) to launch.
+        Each worker will reserve 1 CPU by default. The number of CPUs
+        reserved by each worker can be overridden with the
+        ``resources_per_worker`` argument.
+    use_gpu: If True, training will be done on GPUs (1 per worker).
+        Defaults to False. The number of GPUs reserved by each
+        worker can be overridden with the ``resources_per_worker``
+        argument.
+    resources_per_worker: If specified, the resources
+        defined in this Dict will be reserved for each worker. The
+        ``CPU`` and ``GPU`` keys (case-sensitive) can be defined to
+        override the number of CPU/GPUs used by each worker.
+    placement_strategy: The placement strategy to use for the
+        placement group of the Ray actors. See :ref:`Placement Group
+        Strategies <pgroup-strategy>` for the possible options.
+    """
+
+    num_workers: int
+    use_gpu: bool = False
+    resources_per_worker: Optional[Dict] = None
+    placement_strategy: str = "PACK"
+
+    @property
+    def num_cpus_per_worker(self):
+        """The number of CPUs to set per worker."""
+        raise NotImplementedError
+
+    @property
+    def num_gpus_per_worker(self):
+        """The number of GPUs to set per worker."""
+        raise NotImplementedError
+
+    @property
+    def additional_resources_per_worker(self):
+        """Resources per worker, not including CPU or GPU resources."""
+        raise NotImplementedError
+
+    def get_placement_group_factory(self) -> "PlacementGroupFactory":
```

Comment:
nit: Would this be more intuitive if named `as_placement_group_factory` or `to_placement_group_factory`? Since this is essentially converting the object to another representation as opposed to retrieving an attribute.

Reply:
Changed